test/boost_ext/CompressedStorageComparison.hpp

```diff
@@ -0,0 +1,107 @@
+/****************************************************************************
+ * Copyright (c) 2012-2020 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_BOOST_TEST_COMPRESSED_STORAGE_COMPARISON_HPP
+#define ARBORX_BOOST_TEST_COMPRESSED_STORAGE_COMPARISON_HPP
+
+#include <boost/test/tools/detail/print_helper.hpp>
+
+#include <iosfwd>
+#include <set>
+
+template <typename Offsets, typename Values>
+struct CompressedStorage
+{
+  Offsets offsets;
+  Values values;
+  using value_type = typename Values::value_type;
+  using index_type = typename Offsets::value_type;
+  struct ConstForwardIterator
+  {
+    index_type i;
+    CompressedStorage const *p;
+    using value_type = std::multiset<CompressedStorage::value_type>;
+    ConstForwardIterator &operator++()
+    {
+      ++i;
+      return *this;
+    }
+    ConstForwardIterator operator++(int)
+    {
+      ConstForwardIterator old{*this};
+      ++*this;
+      return old;
+    };
+    ConstForwardIterator operator==(ConstForwardIterator const &o)
+    {
+      return i == o.i;
+    }
+    ConstForwardIterator operator!=(ConstForwardIterator const &o)
+    {
+      return !(*this == o);
+    }
+    value_type operator*()
+    {
+      return {p->values.data() + *(p->offsets.data() + i),
+              p->values.data() + *(p->offsets.data() + i + 1)};
```

Comment:
These are problematic, and could lead to a segfault. Need to check that we don't get beyond `p->values` size. Also, is there a reason to write `*(p->offsets.data() + i+1)` instead of `p->offsets[i+1]`? That should work for both `std::vector` and `Kokkos::View`.

Reply:
I forgot about `View::operator[]`.  It does not protect you against out of bounds access either though :p

Reply:
I believe that we have `Kokkos_ENABLE_DEBUG_BOUNDS_CHECK`  for `View::operator[]`.